ledger/ledger.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ledger
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+type Ledger interface {
+	Put(key, value string) (string, error)
+	Delete(key string) error
+	Get(key string) (string, error)
+	RootHash() string
+	GetPreviousValue(previousHash, key string) (result string, err error)
+}
+
+type SMTLedger struct {
+	tree SMT
+}
+
+// Make returns a Ledger which will retain previous nodes after they are deleted.
+func Make(retention time.Duration) Ledger {
+	return &SMTLedger{tree: *NewSMT(nil, Hasher, nil, retention)}
+}
+
+// Put adds a key value pair to the ledger, overwriting previous values and marking them for
+// removal after the retention specified in Make()
+func (s *SMTLedger) Put(key, value string) (result string, err error) {
+	b, err := s.tree.Update([][]byte{[]byte(fmt.Sprintf("%8v", key))},
```

Comment:
Is this code perf-sensitive? Cause Sprintf is a fairly slow function, which induces garbage to be created. Why is Sprintf needed instead of just using the input strings directly?

Reply:
I was not aware that Sprintf was not performant.  I'm just trying to enforce that the key is 8 bytes in length, with a sane zerofill in shorter keys and truncation in larger ones... I will find a way to rework this with more performant code.